src/ekat/ekat_physical_constants.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef EKAT_PHYSICAL_CONSTANTS
+#define EKAT_PHYSICAL_CONSTANTS
+
+#include "ekat/util/ekat_units.hpp"
+
+namespace ekat {
+
+template <typename RealType=double>
+class PhysicalConstant {
+  public:
+    //enum Op {Times, Divide};
+
+    // no default
+    PhysicalConstant() = delete;
+
+    PhysicalConstant(const PhysicalConstant<RealType>&) = default;
+
+    // Constructor for single value (i.e., NIST std)
+    PhysicalConstant(const RealType& nvalue, const ekat::units::Units& units, const std::string& name) :
+      m_nist_val(nvalue), m_e3sm_val(nvalue), m_units(units), m_name(name), m_info("") {}
+
+    // Constructor for single value (i.e., NIST std)
+    PhysicalConstant(const RealType& nvalue, const ekat::units::Units& units, const std::string& name,
+      const std::string& info) :
+      m_nist_val(nvalue), m_e3sm_val(nvalue), m_units(units), m_name(name), m_info(info) {}
+
+
+    // Constructor for two different values (i.e., when E3SM value and NIST value are different)
+    PhysicalConstant(const RealType& evalue, const RealType& nvalue, const ekat::units::Units& units,
+      const std::string& name) : m_e3sm_val(evalue), m_nist_val(nvalue), m_units(units),
+      m_name(name), m_info("") {}
+
+    // Constructor for two different values (i.e., when E3SM value and NIST value are different)
+    PhysicalConstant(const RealType& evalue, const RealType& nvalue, const ekat::units::Units& units,
+      const std::string& name, const std::string& info) : m_e3sm_val(evalue), m_nist_val(nvalue), m_units(units),
+      m_name(name), m_info(info) {}
+
+    // Constructor from the product of two other PhysicalConstants
+    PhysicalConstant(const PhysicalConstant& a, const PhysicalConstant& b, const std::string& name,
+      const std::string& info="") : m_nist_val(a.m_nist_val*b.m_nist_val),
+        m_e3sm_val(a.m_e3sm_val*b.m_e3sm_val), m_units(a.m_units*b.m_units),
+        m_name(name), m_info(info) {}
+
+    std::string name() const {return m_name;}
+
+    std::string get_string() const {return m_name + " [" + to_string(m_units) +
+      (m_info.empty() ? "]" : "] (" + m_info + ")");}
+
+    std::string unit_string() const {return to_string(m_units);}
+
+    RealType value() const {return m_e3sm_val;}
+
+    RealType operator() () const {return value();}
```

Comment:
I'm curious what others think about using an implicit casting operator for `RealType` here, which would eliminate a set of parenthesis in our math expressions. Given that constants need to be converted to packs (using `sp(constant)`, for example, in SCREAM), maybe that would look a little cleaner. (I'm not usually an advocate for implicit casting, mind you.)

And maybe it would be a good idea to take some example expressions from P3 and SHOC in SCREAM and see how this class might be used to simplify them, or at least see how it looks when used in that context.

Reply:
Agree.  Once I get a basic design ok'd, finding a few examples in p3 and shoc is where I'll go next.  I won't put a final PR in until we are ok with both the design and how it's used.

Reply:
I think implicit cast is ok here. One is already specifying the underlying precision via the RealType, so there's not intrinsic risk of unexpected precision loss/gain.